File: typescript/packages/core/test/integrations/core.test.ts
import { beforeEach, describe, expect, it } from "vitest";
import { x402Client, x402HTTPClient } from "../../src/client";
import { x402Facilitator } from "../../src/facilitator";
import {
  HTTPAdapter,
  HTTPResponseInstructions,
  x402HTTPResourceService,
  x402ResourceService,
} from "../../src/server";
import {
  buildCashPaymentRequirements,
  CashFacilitatorClient,
  CashSchemeNetworkClient,
  CashSchemeNetworkFacilitator,
  CashSchemeNetworkService,
} from "./mocks/cash";
import { Network, PaymentPayload, PaymentRequirements } from "../../src/types";

describe("Core Integration Tests", () => {
  describe("x402Client / x402ResourceService / x402Facilitator - Cash Flow", () => {
    let client: x402Client;
    let server: x402ResourceService;

    beforeEach(async () => {
      client = new x402Client().registerScheme("x402:cash", new CashSchemeNetworkClient("John"));

      const facilitator = new x402Facilitator().registerScheme(
        "x402:cash",
        new CashSchemeNetworkFacilitator(),
      );

      const facilitatorClient = new CashFacilitatorClient(facilitator);
      server = new x402ResourceService(facilitatorClient);
      server.registerScheme("x402:cash", new CashSchemeNetworkService());
      await server.initialize(); // Initialize to fetch supported kinds
    });

    it("server should successfully verify and settle a cash payment from a client", async () => {
      // Server - builds PaymentRequired response
      const accepts = [buildCashPaymentRequirements("Company Co.", "USD", "1")];
      const resource = {
        url: "https://company.co",
        description: "Company Co. resource",
        mimeType: "application/json",
      };
      const paymentRequiredResponse = server.createPaymentRequiredResponse(accepts, resource);

      // Client - responds with PaymentPayload response
      const selected = client.selectPaymentRequirements(
        paymentRequiredResponse.x402Version,
        accepts,
      );
      const paymentPayload = await client.createPaymentPayload(
        paymentRequiredResponse.x402Version,
        selected,
      );

      // Server - maps payment payload to payment requirements
      const accepted = server.findMatchingRequirements(accepts, paymentPayload);
      expect(accepted).toBeDefined();

      const verifyResponse = await server.verifyPayment(paymentPayload, accepted!);
      expect(verifyResponse.isValid).toBe(true);

      // Server does work here

      const settleResponse = await server.settlePayment(paymentPayload, accepted!);
      expect(settleResponse.success).toBe(true);
    });
  });

  describe("x402HTTPClient / x402HTTPResourceService / x402Facilitator - Cash Flow", () => {
    let client: x402HTTPClient;
    let service: x402HTTPResourceService;

    const routes = {
      "/api/protected": {
        scheme: "cash",
        payTo: "merchant@example.com",
        price: "$0.10",
        network: "x402:cash" as Network,
        description: "Access to protected API",
        mimeType: "application/json",
      },
    };

    const mockAdapter: HTTPAdapter = {
      getHeader: (name: string) => {
        // Return payment header if requested
        if (name === "x-payment") {
          return "base64EncodedPaymentHere";
        }
        return undefined;
      },
      getMethod: () => "GET",
      getPath: () => "/api/protected",
      getUrl: () => "https://example.com/api/protected",
      getAcceptHeader: () => "application/json",
      getUserAgent: () => "TestClient/1.0",
    };

    beforeEach(async () => {
      const facilitator = new x402Facilitator().registerScheme(
        "x402:cash",
        new CashSchemeNetworkFacilitator(),
      );

      const facilitatorClient = new CashFacilitatorClient(facilitator);

      client = new x402HTTPClient().registerScheme(
        "x402:cash",
        new CashSchemeNetworkClient("John"),
      ) as x402HTTPClient;

      service = new x402HTTPResourceService(routes, facilitatorClient);
      service.registerScheme("x402:cash", new CashSchemeNetworkService());
      await service.initialize(); // Initialize to fetch supported kinds
    });

    it("middleware should successfully verify and settle a cash payment from an http client", async () => {
      // Middleware creates a PaymentRequired response
      const context = {
        adapter: mockAdapter,
        path: "/api/protected",
        method: "GET",
      };
      // No payment made, get PaymentRequired response & header
      const httpProcessResult = (await service.processHTTPRequest(context))!;

      expect(httpProcessResult.type).toBe("payment-error");

      const initial402Response = (
        httpProcessResult as { type: "payment-error"; response: HTTPResponseInstructions }
      ).response;

      expect(initial402Response).toBeDefined();
      expect(initial402Response.status).toBe(402);
      expect(initial402Response.headers).toBeDefined();
      expect(initial402Response.headers["PAYMENT-REQUIRED"]).toBeDefined();
      expect(initial402Response.isHtml).toBeFalsy();
      expect(initial402Response.body).toBeUndefined();

      // Client responds to PaymentRequired and submits a request with a PaymentPayload
      const paymentRequired = client.getPaymentRequiredResponse(
        initial402Response.headers,
        initial402Response.body,
      );
      const selected = client.selectPaymentRequirements(
        paymentRequired.x402Version,
        paymentRequired.accepts,
      );
      const paymentPayload = await client.createPaymentPayload(
        paymentRequired.x402Version,
        selected,
      );
      const requestHeaders = await client.encodePaymentSignatureHeader(paymentPayload);

      // Middleware handles PAYMENT-SIGNATURE request
      context.adapter.getHeader = (name: string) => {
        if (name === "PAYMENT-SIGNATURE") {
          return requestHeaders["PAYMENT-SIGNATURE"];
        }
        return undefined;
      };
      const httpProcessResult2 = await service.processHTTPRequest(context);

      // No need to reason respond, can continue with request
      expect(httpProcessResult2.type).toBe("payment-verified");
      const {
        paymentPayload: verifiedPaymentPayload,
        paymentRequirements: verifiedPaymentRequirements,
      } = httpProcessResult2 as {
        type: "payment-verified";
        paymentPayload: PaymentPayload;
        paymentRequirements: PaymentRequirements;
      };

      const settlementHeaders = await service.processSettlement(
        verifiedPaymentPayload,
        verifiedPaymentRequirements,
        200,
      );
      expect(settlementHeaders).toBeDefined();
      expect(settlementHeaders?.["PAYMENT-RESPONSE"]).toBeDefined();
    });
  });
});

File: go/test/integration/core_test.go
package integration_test

import (
	"context"
	"encoding/json"
	"testing"

	x402 "github.com/coinbase/x402/go"
	"github.com/coinbase/x402/go/test/mocks/cash"
)

// TestCoreIntegration tests the integration between x402Client, x402ResourceServer, and x402Facilitator
func TestCoreIntegration(t *testing.T) {
	t.Run("Cash Flow - x402Client / x402ResourceServer / x402Facilitator", func(t *testing.T) {
		ctx := context.Background()

		// Setup client with cash scheme
		client := x402.Newx402Client()
		client.RegisterScheme("x402:cash", cash.NewSchemeNetworkClient("John"))

		// Setup facilitator with cash scheme
		facilitator := x402.Newx402Facilitator()
		facilitator.RegisterScheme("x402:cash", cash.NewSchemeNetworkFacilitator())

		// Create facilitator client wrapper
		facilitatorClient := cash.NewFacilitatorClient(facilitator)

		// Setup resource server
		server := x402.Newx402ResourceServer(
			x402.WithFacilitatorClient(facilitatorClient),
		)
		server.RegisterScheme("x402:cash", cash.NewSchemeNetworkServer())

		// Initialize server to fetch supported kinds
		err := server.Initialize(ctx)
		if err != nil {
			t.Fatalf("Failed to initialize server: %v", err)
		}

		// Server - builds PaymentRequired response
		accepts := []x402.PaymentRequirements{
			cash.BuildPaymentRequirements("Company Co.", "USD", "1"),
		}
		resource := x402.ResourceInfo{
			URL:         "https://company.co",
			Description: "Company Co. resource",
			MimeType:    "application/json",
		}
		paymentRequiredResponse := server.CreatePaymentRequiredResponse(accepts, resource, "", nil)

		// Client - responds with PaymentPayload response
		selected, err := client.SelectPaymentRequirements(paymentRequiredResponse.X402Version, accepts)
		if err != nil {
			t.Fatalf("Failed to select payment requirements: %v", err)
		}

		// Marshal selected requirements to bytes
		selectedBytes, err := json.Marshal(selected)
		if err != nil {
			t.Fatalf("Failed to marshal requirements: %v", err)
		}

		payloadBytes, err := client.CreatePaymentPayload(ctx, paymentRequiredResponse.X402Version, selectedBytes, nil, nil)
		if err != nil {
			t.Fatalf("Failed to create payment payload: %v", err)
		}

		// Server - maps payment payload to payment requirements
		accepted := server.FindMatchingRequirements(accepts, payloadBytes)
		if accepted == nil {
			t.Fatal("No matching payment requirements found")
		}

		// Marshal accepted requirements to bytes
		acceptedBytes, err := json.Marshal(accepted)
		if err != nil {
			t.Fatalf("Failed to marshal accepted requirements: %v", err)
		}

		// Server - verifies payment
		verifyResponse, err := server.VerifyPayment(ctx, payloadBytes, acceptedBytes)
		if err != nil {
			t.Fatalf("Failed to verify payment: %v", err)
		}

		if !verifyResponse.IsValid {
			t.Fatalf("Payment verification failed: %s", verifyResponse.InvalidReason)
		}

		// Server does work here...

		// Server - settles payment
		settleResponse, err := server.SettlePayment(ctx, payloadBytes, acceptedBytes)
		if err != nil {
			t.Fatalf("Failed to settle payment: %v", err)
		}

		if !settleResponse.Success {
			t.Fatalf("Payment settlement failed: %s", settleResponse.ErrorReason)
		}

		// Verify the transaction message
		expectedTransaction := "John transferred 1 USD to Company Co."
		if settleResponse.Transaction != expectedTransaction {
			t.Errorf("Expected transaction '%s', got '%s'", expectedTransaction, settleResponse.Transaction)
		}
	})
}
